### PR TITLE
Increase contributor count

### PR DESCRIPTION
**Pull request by darcys22**

### CODE REVIEW

**Comment by jagerman** [comment]
I think we are going to have to shelve the staking portions == staking requirement part of this.  Looking at the code, there are definite issues present such as when we accept the different registrations, and how we interpret fees on the blockchain.  For instance, if we have a service node record with a raw fee value set to 3750000000000 then how do we know whether that means 25% (base 15e12) or 0.00002% (base 2^64-4)?  The tests here basically change everything to the former, but in reality we will have both types of registrations on the network that we have to deal with.

That in turn is going to mean that we have to have to introduce some new version of the registration tx that indicates the registration version (and thus the base), which we have to hardfork-version-gate all over the place, and have to chase through all the calculations of fees, etc. to be aware of the registration type so that they can figure out which base they are working with.

While I'd still like to do it (because the base-2^64-4 stuff is painful), I think we need to shelve it for now and pursue it for the next hard fork.

File: src/cryptonote_config.h [hunk 1]
```diff
@@ -50,12 +50,17 @@ using namespace std::literals;
...
 
+#define STAKING_REQUIREMENT_HF16                        15000000000000
 #define STAKING_REQUIREMENT_LOCK_BLOCKS_EXCESS          20
-#define STAKING_PORTIONS                                UINT64_C(0xfffffffffffffffc)
-#define MAX_NUMBER_OF_CONTRIBUTORS                      4
-#define MIN_PORTIONS                                    (STAKING_PORTIONS / MAX_NUMBER_OF_CONTRIBUTORS)
-
-static_assert(STAKING_PORTIONS % 12 == 0, "Use a multiple of twelve, so that it divides evenly by two, three, or four contributors.");
+#define STAKING_PORTIONS_V1                             UINT64_C(0xfffffffffffffffc)
+#define STAKING_PORTIONS_V2                             STAKING_REQUIREMENT_HF16
```
Comment (jagerman) [inline src/cryptonote_config.h:56]: This needs to stay at the old value.  (Don't just revert to the previous 0xfffffffffffffff0 change in this PR).  If we change this base at all there is a good risk of introducing subtle changes into the calculations of rewards that could lead to consensus disagreements: in between the release and the hard fork we'll have some nodes calculating using 0xfff...c and some calculating using 0xfff...0 -- and if that introduces an off-by-one error we could have some subset of nodes that accept a registration and another that doesn't.

File: src/cryptonote_config.h [hunk 2]
```diff
@@ -50,12 +50,17 @@ using namespace std::literals;
...
+#define MAX_NUMBER_OF_CONTRIBUTORS_V1                   4
+#define MAX_NUMBER_OF_CONTRIBUTORS_V2                   10
+#define MIN_PORTIONS_V2                                 (STAKING_PORTIONS_V2 / MAX_NUMBER_OF_CONTRIBUTORS_V2)
+#define MIN_PORTIONS_V1                                 (STAKING_PORTIONS_V1 / MAX_NUMBER_OF_CONTRIBUTORS_V1)
+#define MIN_OPERATOR_PORTION                            (STAKING_PORTIONS_V2 / MAX_NUMBER_OF_CONTRIBUTORS_V1)
```
Comment (jagerman) [inline src/cryptonote_config.h:61]: MIN_OPERATOR_PORTION is obviously still needed, but is this MIN_PORTIONS_V2 needed at all?  It seems to me that everyone after the operator is entirely dependent on what was staked before, which makes `MIN_PORTIONS` irrelevant.  (Though there might be some need for the old MIN_PORTIONS_V1 value to stick around for older HF chain handling).

File: src/cryptonote_core/service_node_list.cpp [hunk 3]
```diff
@@ -337,9 +337,9 @@ namespace service_nodes
       throw invalid_contributions{"No portions given"};
     if (contributor_args.portions.size() != contributor_args.addresses.size())
       throw invalid_contributions{"Number of portions (" + std::to_string(contributor_args.portions.size()) + ") doesn't match the number of addresses (" + std::to_string(contributor_args.portions.size()) + ")"};
-    if (contributor_args.portions.size() > MAX_NUMBER_OF_CONTRIBUTORS)
+    if (contributor_args.portions.size() > MAX_NUMBER_OF_CONTRIBUTORS_V2)
```
Comment (jagerman) [inline src/cryptonote_core/service_node_list.cpp:340]: Needs to be HF-aware (i.e. we can't allow a 5-contributor node before HF19).

File: src/cryptonote_config.h (hunk 1 above)
Comment (jagerman) [reply src/cryptonote_config.h:56]: (See my other comment about putting off the portions == requirement change for now; when we reintroduce that, we'll need a bunch of hardfork gating around which value gets used where).

File: src/cryptonote_config.h (hunk 2 above)
Comment (majestrate) [reply src/cryptonote_config.h:61]: is there any reason these are macros besides "tradition"?

File: src/cryptonote_core/service_node_rules.cpp [hunk 4]
```diff
@@ -89,23 +89,23 @@ uint64_t portions_to_amount(uint64_t portions, uint64_t staking_requirement)
...
   lo = mul128(staking_requirement, portions, &hi);
-  div128_64(hi, lo, STAKING_PORTIONS, &resulthi, &resultlo);
+  div128_64(hi, lo, STAKING_PORTIONS_V2, &resulthi, &resultlo);
   return resultlo;
 }
 
 bool check_service_node_portions(uint8_t hf_version, const std::vector<uint64_t>& portions)
 {
-  if (portions.size() > MAX_NUMBER_OF_CONTRIBUTORS) return false;
+  if (portions.size() > MAX_NUMBER_OF_CONTRIBUTORS_V2) return false;
```
Comment (jagerman) [inline src/cryptonote_core/service_node_rules.cpp:98]: Depends on `hf_version`

File: src/cryptonote_basic/hardfork.cpp [hunk 5]
```diff
@@ -75,10 +75,10 @@ static constexpr std::array devnet_hard_forks =
...
   hard_fork{ 13, 0,     4,  1599848400 }, 
   hard_fork{ 15, 0,     5,  1599848400 },
-  hard_fork{ 16, 0,   100,  1599848400 },
-  hard_fork{ 17, 0,   151,  1599848400 },
-  hard_fork{ 18, 0,   152,  1599848400 },
-  hard_fork{ 19, 0,   153,  1599848400 },
+  hard_fork{ 16, 0,   199,  1599848400 },
+  hard_fork{ 17, 0,   251,  1599848400 },
+  hard_fork{ 18, 0,   252,  1599848400 },
+  hard_fork{ 19, 0,   253,  1599848400 },
```
Comment (jagerman) [inline src/cryptonote_basic/hardfork.cpp:81]: Is this change necessary?  This is going to hard break current devnet.

File: src/cryptonote_core/service_node_list.cpp [hunk 6]
```diff
@@ -994,6 +994,12 @@ namespace service_nodes
       return false;
     }
 
+    if (hf_version >= cryptonote::network_version_19 && stake.transferred < get_min_node_operator_contribution(info.staking_requirement))
+    {
+      LOG_PRINT_L1("Register TX rejected: TX does not have sufficient operator stake");
+      return false;
+    }
+
```
Comment (jagerman) [inline src/cryptonote_core/service_node_list.cpp:1002]: This check is in the wrong place: it should be done as part of service_node_rules.cpp's `check_service_node_portions(...)` where we check all the other portions.  For hf >= 19 it can special-case the `i == 0` case to use `get_min_node_operator_contribution` rather than `get_min_node_contribution` and check all contribution requirements in one go/one place.  (And then once that check is there, this one is unnecessary because we'll have already ensured that `reserved >= required`, and the previous `if` statement assures `transfered >= reserved`.

File: src/cryptonote_core/service_node_list.cpp [hunk 7]
```diff
@@ -1145,20 +1151,23 @@ namespace service_nodes
     // Check node contributor counts
     {
       bool too_many_contributions = false;
-      if (hf_version >= cryptonote::network_version_16_pulse)
+      if (hf_version >= cryptonote::network_version_19)
+        // As of HF19 we allow up to 10 stakes total (except for the loophole closed above)
```
Comment (jagerman) [inline src/cryptonote_core/service_node_list.cpp:1155]:
```suggestion
        // As of HF19 we allow up to 10 stakes total
```

File: src/cryptonote_core/service_node_list.cpp [hunk 8]
```diff
@@ -1145,20 +1151,23 @@ namespace service_nodes
...
       else
         // Before HF11 we allowed up to 4 contributors, but each can contribute multiple times
-        too_many_contributions = new_contributor && contributors.size() >= MAX_NUMBER_OF_CONTRIBUTORS;
+        too_many_contributions = new_contributor && contributors.size() >= MAX_NUMBER_OF_CONTRIBUTORS_V1;
 
       if (too_many_contributions)
       {
         LOG_PRINT_L1("TX: Already hit the max number of contributions: "
-                     << MAX_NUMBER_OF_CONTRIBUTORS
+                     << MAX_NUMBER_OF_CONTRIBUTORS_V2
```
Comment (jagerman) [inline src/cryptonote_core/service_node_list.cpp:1170]:
```suggestion
                     << (hf_version >= 19 ? MAX_NUMBER_OF_CONTRIBUTORS_V2 : MAX_NUMBER_OF_CONTRIBUTORS_V1)
```

File: src/cryptonote_core/service_node_list.cpp [hunk 9]
```diff
@@ -3665,24 +3674,24 @@ namespace service_nodes
       return result;
     }
 
-    if ((args.size()-1)/ 2 > MAX_NUMBER_OF_CONTRIBUTORS)
+    if ((args.size()-1)/ 2 > MAX_NUMBER_OF_CONTRIBUTORS_V2)
```
Comment (jagerman) [inline src/cryptonote_core/service_node_list.cpp:3677]:
```suggestion
    const size_t max_contributors = hf_version >= 19 ? MAX_NUMBER_OF_CONTRIBUTORS_V2 : MAX_NUMBER_OF_CONTRIBUTORS;
    if (args.size() > 1 + 2 * max_contributors)
```

File: src/cryptonote_core/service_node_list.cpp [hunk 10]
```diff
@@ -3665,24 +3674,24 @@ namespace service_nodes
...
     {
-      result.err_msg = tr("Exceeds the maximum number of contributors, which is ") + std::to_string(MAX_NUMBER_OF_CONTRIBUTORS);
+      result.err_msg = tr("Exceeds the maximum number of contributors, which is ") + std::to_string(MAX_NUMBER_OF_CONTRIBUTORS_V2);
```
Comment (jagerman) [inline src/cryptonote_core/service_node_list.cpp:3679]:
```suggestion
      result.err_msg = tr("Exceeds the maximum number of contributors, which is ") + std::to_string(max_contributors);
```

File: src/cryptonote_core/service_node_list.cpp [hunk 11]
```diff
@@ -3793,19 +3803,19 @@ namespace service_nodes
...
           // the minimum by dust, just increase the portions so it passes
-          if (needed > 0 && addr_to_portions.size() < MAX_NUMBER_OF_CONTRIBUTORS)
+          if (needed > 0 && addr_to_portions.size() < MAX_NUMBER_OF_CONTRIBUTORS_V2)
             addr_to_portion.portions += needed;
       }
 
       if (addr_to_portion.portions < min_portions || (addr_to_portion.portions - portions_to_steal) > portions_left)
       {
-        result.err_msg = tr("Invalid amount for contributor: ") + args[i] + tr(", with portion amount: ") + args[i+1] + tr(". The contributors must each have at least 25%, except for the last contributor which may have the remaining amount");
+        result.err_msg = tr("Invalid amount for contributor: ") + args[i] + tr(", with portion amount: ") + args[i+1] + tr(". The contributors must each have at least 25%, except for the last contributor which may have the remaining         amount");
```
Comment (jagerman) [inline src/cryptonote_core/service_node_list.cpp:3812]: Spaces added by mistake?

File: src/cryptonote_core/service_node_list.cpp [hunk 12]
```diff
@@ -3793,19 +3803,19 @@ namespace service_nodes
...
         return result;
       }
 
       if (min_portions == UINT64_MAX)
       {
-        result.err_msg = tr("Too many contributors specified, you can only split a node with up to: ") + std::to_string(MAX_NUMBER_OF_CONTRIBUTORS) + tr(" people.");
+        result.err_msg = tr("Too many contributors specified, you can only split a node with up to: ") + std::to_string(MAX_NUMBER_OF_CONTRIBUTORS_V2) + tr(" people.");
```
Comment (jagerman) [inline src/cryptonote_core/service_node_list.cpp:3818]:
```suggestion
        result.err_msg = tr("Too many contributors specified, you can only split a node with up to: ") + std::to_string(max_contributors) + tr(" people.");
```

File: src/cryptonote_config.h (hunk 2 above)
Comment (jagerman) [reply src/cryptonote_config.h:61]: > is there any reason these are macros besides "tradition"?

Nope.  There's a `config` namespace in here with constexpr values, which is probably a good place to move all of these.

Comment (majestrate) [reply src/cryptonote_config.h:61]: that'd probably spiral into a full refactor of everything around it, good idea to omit this task for now.

File: src/cryptonote_basic/hardfork.cpp (hunk 5 above)
Comment (darcys22) [reply src/cryptonote_basic/hardfork.cpp:81]: not necessary but I made the local devnet scripts send a bunch of funds to new wallets for a node with 10 contributors so it needed extra funds. Will remove

**Comment by darcys22** [comment]
Updates made, tests are passing for this branch but depends on batching PR which needs to be merged first

File: src/cryptonote_core/service_node_list.cpp [hunk 13]
```diff
@@ -1145,20 +1146,23 @@ namespace service_nodes
     // Check node contributor counts
     {
       bool too_many_contributions = false;
-      if (hf_version >= cryptonote::network_version_16_pulse)
+      if (hf_version >= cryptonote::network_version_19)
+        // As of HF19 we allow up to 10 stakes total
+        too_many_contributions = existing_contributions + stake.locked_contributions.size() > MAX_NUMBER_OF_CONTRIBUTORS_V2;
```
Comment (jagerman) [inline src/cryptonote_core/service_node_list.cpp:1151]: Should this not be:
```suggestion
        too_many_contributions = existing_contributions + other_reservations + 1 > MAX_NUMBER_OF_CONTRIBUTORS_V2;
```
to match the logic under HF16 (but just changing the RHS variable)?

**Comment by jagerman** [comment]
One change here that doesn't seem quite right, but otherwise this looks good to go.

File: src/cryptonote_core/service_node_rules.cpp [hunk 14]
```diff
@@ -158,6 +169,11 @@ uint64_t get_min_node_contribution_in_portions(uint8_t version, uint64_t staking
   return result;
 }
 
+uint64_t get_min_node_operator_contribution(uint64_t staking_requirement)
+{
+  return staking_requirement / MAX_NUMBER_OF_CONTRIBUTORS_V1;
```
Comment (jagerman) [inline src/cryptonote_core/service_node_rules.cpp:174]: This calculation feel wrong to me, though the actual *value* it calculates isn't.  It just feels wrong to be depending on a `_V1` constant when we aren't at V1 anymore.  I think instead we should have a new constant for this, something like
`MIN_OPERATOR_SHARE` that is set to 4.
